### ISSUE TITLE
newdtype example doesn't work with python3

### ISSUE THREAD

**Issue by modkzs** [numpy/numpy#9282]
python 3 change some c api, so newdtype example in doc can't compile

**Comment by ahaldane** [comment]
It's not clear what you are referring to. Could you paste and example illustrating the problem?

**Comment by modkzs** [comment]
I try to compile newdtype_example but get errors:
```
/Users/megvii/.pyenv/versions/3.5.0/lib/python3.5/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:15:2: warning: "Using deprecated NumPy API, disable it by "          "#defining NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-W#warnings]
#warning "Using deprecated NumPy API, disable it by " \
 ^
floatint.c:17:5: warning: suggest braces around initialization of subobject [-Wmissing-braces]
    PyObject_HEAD_INIT(NULL)
    ^~~~~~~~~~~~~~~~~~~~~~~~
/Users/megvii/.pyenv/versions/3.5.0/include/python3.5m/object.h:87:5: note: expanded from macro 'PyObject_HEAD_INIT'
    1, type },
    ^~~~~~~
floatint.c:19:5: warning: incompatible pointer to integer conversion initializing 'Py_ssize_t' (aka 'long') with an expression of type 'char [18]' [-Wint-conversion]
    "floatint.floatint",                        /*tp_name*/
    ^~~~~~~~~~~~~~~~~~~
floatint.c:105:23: error: no member named 'ob_type' in 'struct _PyArray_Descr'
    _PyFloatInt_Dtype.ob_type = &PyArrayDescr_Type;
    ~~~~~~~~~~~~~~~~~ ^
floatint.c:119:9: warning: implicit declaration of function 'Py_InitModule' is invalid in C99 [-Wimplicit-function-declaration]
    m = Py_InitModule("floatint", NULL);
        ^
floatint.c:119:7: warning: incompatible integer to pointer conversion assigning to 'PyObject *' (aka 'struct _object *') from 'int' [-Wint-conversion]
    m = Py_InitModule("floatint", NULL);
      ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
floatint.c:130:42: error: non-void function 'initfloatint' should return a value [-Wreturn-type]
    if (PyType_Ready(&PyFloat_Type) < 0) return;
                                         ^
floatint.c:140:45: error: non-void function 'initfloatint' should return a value [-Wreturn-type]
    if (PyType_Ready(&PyFloatInt_Type) < 0) return;
                                            ^
floatint.c:151:5: error: non-void function 'initfloatint' should return a value [-Wreturn-type]
    return;
    ^
5 warnings and 4 errors generated.
In file included from floatint.c:4:
In file included from /Users/megvii/.pyenv/versions/3.5.0/lib/python3.5/site-packages/numpy/core/include/numpy/arrayobject.h:4:
In file included from /Users/megvii/.pyenv/versions/3.5.0/lib/python3.5/site-packages/numpy/core/include/numpy/ndarrayobject.h:18:
In file included from /Users/megvii/.pyenv/versions/3.5.0/lib/python3.5/site-packages/numpy/core/include/numpy/ndarraytypes.h:1809:
/Users/megvii/.pyenv/versions/3.5.0/lib/python3.5/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:15:2: warning: "Using deprecated NumPy API, disable it by "          "#defining NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-W#warnings]
#warning "Using deprecated NumPy API, disable it by " \
 ^
floatint.c:17:5: warning: suggest braces around initialization of subobject [-Wmissing-braces]
    PyObject_HEAD_INIT(NULL)
    ^~~~~~~~~~~~~~~~~~~~~~~~
/Users/megvii/.pyenv/versions/3.5.0/include/python3.5m/object.h:87:5: note: expanded from macro 'PyObject_HEAD_INIT'
    1, type },
    ^~~~~~~
floatint.c:19:5: warning: incompatible pointer to integer conversion initializing 'Py_ssize_t' (aka 'long') with an expression of type 'char [18]' [-Wint-conversion]
    "floatint.floatint",                        /*tp_name*/
    ^~~~~~~~~~~~~~~~~~~
floatint.c:105:23: error: no member named 'ob_type' in 'struct _PyArray_Descr'
    _PyFloatInt_Dtype.ob_type = &PyArrayDescr_Type;
    ~~~~~~~~~~~~~~~~~ ^
floatint.c:119:9: warning: implicit declaration of function 'Py_InitModule' is invalid in C99 [-Wimplicit-function-declaration]
    m = Py_InitModule("floatint", NULL);
        ^
floatint.c:119:7: warning: incompatible integer to pointer conversion assigning to 'PyObject *' (aka 'struct _object *') from 'int' [-Wint-conversion]
    m = Py_InitModule("floatint", NULL);
      ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
floatint.c:130:42: error: non-void function 'initfloatint' should return a value [-Wreturn-type]
    if (PyType_Ready(&PyFloat_Type) < 0) return;
                                         ^
floatint.c:140:45: error: non-void function 'initfloatint' should return a value [-Wreturn-type]
    if (PyType_Ready(&PyFloatInt_Type) < 0) return;
                                            ^
floatint.c:151:5: error: non-void function 'initfloatint' should return a value [-Wreturn-type]
    return;
    ^
5 warnings and 4 errors generated.
```

It seems that python3 changes its c api, so this code doesn't work

**Comment by ahaldane** [comment]
Thanks @modkzs. I confirm the example in `/doc/newdtype_example` doesn't compile in python3, and I don't think it ever has. (The C-api warning is unrelated).

A working custom dtype can be found in `numpy/core/src/umath/test_rational.c.src`, by the way. That file uses some precompiler macros in order to compile in both python2 and python3, which we should copy into the newdtype example.

**Comment by WarrenWeckesser** [comment]
Does the code in `newdtype_example` still serve a useful purpose as documentation?  If not, can we simply remove it?

I've been experimenting with custom dtypes, so if this example is still useful, I would be happy update it, but at the moment, I don't quite see its pedagogical point.   Why create a dtype composed of two 32 bit integers that is a sub-type of a float?  Why are the two integers called `first` and `last`?  Those names suggest some intent, but I don't get it.  I think the example needs some higher-level documentation about goal of this new dtype before I can make sense of the implementation.  Can anyone shed some light on this example?

**Comment by seberg** [comment]
If you cannot see any use in it, and it is buggy, I guess we might as well delete it...

**Comment by WarrenWeckesser** [comment]
Well, I don't know if it is buggy.  It is currently Python 2 only, and I haven't tried to build it with Python 2.  I started to update it for Python 3, but I realized that it would be easier to know if I was doing it correctly if I had a better understanding of the *point* of the example.

It looks like it subtypes a Python float, but items are assigned by passing two integers instead of a float (a bit like a C union of a `double` and two 32 bit integers).
